### PR TITLE
Introduce libgit2

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

This makes most of `GitInputScheme` use `libgit2` instead of the "git" binary. Extracted from the lazy-trees branch, where `GitInputAccessor` is used to provide direct access to git repos without copying them to the Nix store first.

Depends on #9239.

# Context

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by lovesegfault** [comment]
> Also, for the moment, we can't get rid of git for fetching via ssh

Just out of curiosity, why?

**Comment by edolstra** [comment]
> Just out of curiosity, why?

Because the version of `libssh2` that `libgit2` uses doesn't support `ssh_config`.

**Comment by knedlsepp** [comment]
Thanks so much for improving the git-submodules handling (especially with lazy-trees in mind)!  :heart:
Just tried this out on a git repo (sadly non-public) that has a `.gitmodules` file referencing submodules relatively to the git server :

```
[submodule "repo1"]
    path = repo1
    url = ../repo1
    branch = main
```

When using `nix flake show git+ssh://servername/repo?submodules=1` stable `nix` works fine, but with this PR I get:

```
error: file:// URL 'file://../repo1' has unexpected authority '..'
```

When using the local clone of the repository I don't have these issues.

**Comment by lovesegfault** [comment]
> Because the version of libssh2 that libgit2 uses doesn't support ssh_config.

Ah, seems like https://github.com/libgit2/libgit2/pull/6617 will solve that in the next release! 🤞

**Comment by Ericson2314** [comment]
I would very much like to land git hashing support first because it will force us to put this code though its paces from the get-go.

**Comment by Ericson2314** [comment]
> > Because the version of libssh2 that libgit2 uses doesn't support ssh_config.
> 
> Ah, seems like [libgit2/libgit2#6617](https://github.com/libgit2/libgit2/pull/6617) will solve that in the next release! 🤞

Perhaps we should bump to `master` ahead of time. It's much nicer to add temporarily second version of `libgit` to Nixpkgs and drop more code, I think.

(c.f. our fork of `lowdown`)

File: src/libfetchers/git-utils.cc
Comment (roberth) [inline src/libfetchers/git-utils.cc:356]: This seems rather important. Should either implement this, or switch to newer libgit2.

File: src/libfetchers/git.cc
Comment (roberth) [inline src/libfetchers/git.cc:300]:
```suggestion
        /* URL of the repo, or its path if isLocal. Never a `file` URL. */
```

File: src/libfetchers/union-input-accessor.cc
Comment (roberth) [inline src/libfetchers/union-input-accessor.cc:16]:
```suggestion
        // FIXME: return dummy parent directories automatically?
```

I think it's better to do it lazily (regardless of the other decision). They're lazy trees after all.

I'd be happy to unconditionally and lazily generate those directories, even if they shadow pre-existing files.
If this were a primop, a user library could be written to add checking for such "conflicts", but avoiding a built-in check is more difficult and less lazy. But as mentioned, I don't think it's even worth checking, so such a user library is just a last resort.

File: src/libfetchers/union-input-accessor.cc
Comment (roberth) [inline src/libfetchers/union-input-accessor.cc:58]: This would be easier to understand if `ss` wasn't reversed and had a name, but maybe `vector` is faster, and this is fairly hot code.

```suggestion
        std::list<std::string> subpathOfMount;
```

```suggestion
        // ss: reverse subpath of mount point
        std::vector<std::string> ss;
```

File: src/libfetchers/git-utils.cc
Comment (edolstra) [reply src/libfetchers/git-utils.cc:356]: This is existing behavior and this comment was just moved from git.cc. So fixing it is beyond the scope of this PR.

**Comment by fricklerhandwerk** [comment]
Triaged in Nix team meeting:

- @Ericson2314 would like to land [Git hashing](https://github.com/NixOS/nix/issues/8919) first 
  - @edolstra that seems to have no overlap
  - @Ericson2314 if we want Git fetching to be tree-hash oriented, we have to align all the internal book keeping
  - @thufschmitt it seems rather backwards to make the small change depend on the large new feature
  - @edolstra we would only have to make Git hashing end-to-end if we actually stored the fetched result under Git hashes, but we currently don't
  - @roberth sounds great but doesn't sound like it has to be blocking each other
- @Ericson2314 worried that if care isn't taken with the Git hashes that I'll have to re-implement the libgit part
  - (some discussion of the trust model for hashes)
  - @edolstra lazy trees currently will clone the entire repo and then resolve subdirs based on tree hashes using `libgit`
    - it can in principle become more lazy in the future, and fetch subdirs on demand
      - @thufschmitt that would be quite cool
  - @Ericson2314 alright, since the whole graph is offline, hash validation can be done locally
- @thufschmitt wonder how to roll put changes like that. it's a big rewrite of fetchers, so it's easy to sneak in regressions
  - making it experimental would be a pain as that would make for a lot duplication
    - @roberth it may be worth it though
    - @edolstra making it experimental could result in no one using it
      - @thufschmitt one could make it opt-out
      - @edolstra you could just revert in that case
      - @thufschmitt reverting is much more costly for users, and could encourage people to stay on old versions for a long time unnecessarily
      - @roberth as opposed to past times we're now able to fix regressions in a reasonable amount of time
    - another option is to run things in parallel, but for git fetching it may be too costly
      - @roberth given we're intending to fix the smudging bug in a backwards-incompatible way that won't work anyway
  - @ericson2314 personally like the way of running things in parallel
    - @edolstra the point of this is to reduce the diff to the lazy trees branch, and we don't want to maintain two versions
      - @thufschmitt we don't want to *maintain* both, just prevent regressions and then drop the old one once it's safe
      - @edolstra similarly signature verification was merged the other day, and keeping both would mean the versions to diverge
      - @thufschmitt as opposed to the ssh vs. ssh-ng situation (where ssh-ng never reached feature parity) we're not at risk of stalling out
- shall we roll it out and risk breakage or run the two versions in parallel until 
  - @ericson2314 hate the old one, love the new one; we may run into people discovering bugs in the old implementation that they rely on
    - must get rid of shelling out at all cost, as fast as possible
  - @roberth should be okay with releasing the new one alone
  - @edolstra want to get rid of the old one fast
  - @tomberek: the question is if the bugs we expect are such that we can fix them in a timely fashion. then we can keep the new code and proceed. if we're concerned that's not possible, the alternative to keep both becomes more appropriate in terms of cost-benefit ratio
    - @roberth: we can expect some hashes to change. that would be an appropriate breakage for an experimental feature
      - @edolstra: what needs to be done in this PR is testing the current smudging behavior to ensure there won't be a regression
- @roberth could do a call for testing when we deem it good enough
  - @edolstra would prefer that
  - @thufschmitt some time ago we had the idea to ask a large consumer such as NixOS or Home Manager to run their code against `master`, we could try that out here
  - @tomberek one could run some percentage of install actions with master
    - @edolstra that is something the Determinate Systems installer tests are doing in fact, they pick versions at random

File: src/libfetchers/cache.hh
Comment (Ericson2314) [inline src/libfetchers/cache.hh:14]: This should be a doc comment on `Cache`? Also good to put `Attrs` in backticks (Doxygen understands some markdown these days).

File: src/libfetchers/cache.hh
Comment (Ericson2314) [inline src/libfetchers/cache.hh:16]:
```suggestion
    /**
```
Needs the double `**` to be treated as a doc comment, ditto below.

File: src/libfetchers/git-utils.hh
Comment (Ericson2314) [inline src/libfetchers/git-utils.hh:51]: What about bare repos?

**Review by roberth — COMMENTED** [review]
No significant issues. LGTM if you could implement/address the suggestions (this review + previous one + John's).

File: src/libfetchers/git-utils.hh
Comment (edolstra) [reply src/libfetchers/git-utils.hh:51]: AFAIK bare repos just work. There is a test in functional/flakes/flakes.sh.

File: src/libfetchers/git.cc
Comment (roberth) [inline src/libfetchers/git.cc:564]: Why is this check needed?
What if the repo was fetched shallowly and then later we need it without shallow?

If it's part of the cache key that would be a solution. I'd consider that a performance bug still, but could be fixed after merge.

**Review by roberth — APPROVED** [review]
Question about shallow+caching. Otherwise LGTM

File: src/libfetchers/git.cc
Comment (DavHau) [reply src/libfetchers/git.cc:564]: There is much likely a bug in this (already before this PR). If:
- a shallow remote is fetched with `shallow = true`
- later the remote unshallows
- another revision is fetched from the same remote with `shallow = false`
...then this error would trigger and the only way to fix it is deleting the nix cache directory.

This bug could only ever trigger on remote repos, as local repos don't get cached via `~/.cache`.
One could argue that this bug is very unlikely to trigger because remote repos are usually never shallow.

---

It seems to me that the only reason this check has to exist is because on shallow clones the `revCount` breaks. If `revCount` was ignored in general and removed from the API, we wouldn't need to care much about repos being shallow or not.

Comment (DavHau) [reply src/libfetchers/git.cc:564]: To cover potential bugs like this, I would like to improve the testing infra. I opened an issue where this can be discussed: https://github.com/NixOS/nix/issues/9388

File: src/libfetchers/git-utils.cc
Comment (DavHau) [inline src/libfetchers/git-utils.cc:384]: This will fail as soon as the requested `rev` is not the latest `rev` of the fetched `ref`, as any other rev cannot be reached with `--depth 1`. Basically as soon as another commit is made on the remote repo this will certainly fail.

Comment (DavHau) [reply src/libfetchers/git-utils.cc:384]: To solve this, instead of fetching `$ref:$ref`, simply fetching `$rev` with `depth=1` is sufficient. But still, the current `shallow`behavior of builtins.fetchGit would be broken for which I just [clarified the documentation here](https://github.com/NixOS/nix/pull/9390).

**Comment by DavHau** [comment]
@Ericson2314 @edolstra You might have overlooked it, therefore pinging you.
This PR breaks backward compat for `builtins.fetchGit`.
As mentioned in [this comment here](https://github.com/NixOS/nix/pull/9240#discussion_r1398409427), the `--depth 1` introduced by this PR leads to a different result than the previous shallow behavior as [documented here](https://github.com/NixOS/nix/pull/9390).

**Pull request merged**
